bot/exts/moderation/modpings.py

```diff
@@ -1,63 +1,55 @@
 
 from bot.bot import Bot
 from bot.constants import Colours, Emojis, Guild, Icons, MODERATION_ROLES, Roles
 from bot.converters import Expiry
-from bot.utils.scheduling import Scheduler
+from bot.utils.persistent_scheduling import PersistentScheduler
 
 log = logging.getLogger(__name__)
 
 
 class ModPings(Cog):
     """Commands for a moderator to turn moderator pings on and off."""
 
-    # RedisCache[discord.Member.id, 'Naïve ISO 8601 string']
-    # The cache's keys are mods who have pings off.
-    # The cache's values are the times when the role should be re-applied to them, stored in ISO format.
-    pings_off_mods = RedisCache()
-
     def __init__(self, bot: Bot):
         self.bot = bot
-        self._role_scheduler = Scheduler(self.__class__.__name__)
+        self._role_scheduler = PersistentScheduler(self.__class__.__name__, self.reapply_role, bot.loop)
 
         self.guild = None
         self.moderators_role = None
 
-        self.reschedule_task = self.bot.loop.create_task(self.reschedule_roles(), name="mod-pings-reschedule")
+        self.reschedule_task = self.bot.loop.create_task(self.normalize_roles(), name="mod-pings-reschedule")
 
-    async def reschedule_roles(self) -> None:
+    async def normalize_roles(self) -> None:
         """Reschedule moderators role re-apply times."""
         await self.bot.wait_until_guild_available()
         self.guild = self.bot.get_guild(Guild.id)
         self.moderators_role = self.guild.get_role(Roles.moderators)
 
+        await self._role_scheduler.wait_until_ready()
+
         mod_team = self.guild.get_role(Roles.mod_team)
         pings_on = self.moderators_role.members
-        pings_off = await self.pings_off_mods.to_dict()
 
         log.trace("Applying the moderators role to the mod team where necessary.")
         for mod in mod_team.members:
             if mod in pings_on:  # Make sure that on-duty mods aren't in the cache.
-                if mod in pings_off:
-                    await self.pings_off_mods.delete(mod.id)
-                continue
+                if mod.id in self._role_scheduler:
+                    await self._role_scheduler.delete(mod.id)
 
             # Keep the role off only for those in the cache.
-            if mod.id not in pings_off:
-                await self.reapply_role(mod)
-            else:
-                expiry = isoparse(pings_off[mod.id]).replace(tzinfo=None)
-                self._role_scheduler.schedule_at(expiry, mod.id, self.reapply_role(mod))
+            elif mod.id not in self._role_scheduler:
+                await self.reapply_role(mod.id)
 
-    async def reapply_role(self, mod: Member) -> None:
+    async def reapply_role(self, mod_id: int) -> None:
         """Reapply the moderator's role to the given moderator."""
-        log.trace(f"Re-applying role to mod with ID {mod.id}.")
+        log.trace(f"Re-applying role to mod with ID {mod_id}.")
+        await self.reschedule_task
```

Comment:
Doesn't this create a deadlock?

The scheduler calls `reapply_role` and it waits for `reschedule_task`. `reschedule_task` calls `normalize_roles`, which waits for the scheduler to be ready. However, the scheduler is not ready until all `reapply_role` are called. That's a cycle, is it not?

Reply:
Yeah, I tried to be smart and tripped on my own fit. Untangled in 76e62b3